### PR TITLE
Added tests

### PR DESCRIPTION
**Pull request by shirish87**
Scott,

This PR adds tests cases for the public API exposed by the `browserstack` module as well as basic tests validating API responses.

It requires setting the environment variables given below.
````
BROWSERSTACK_USERNAME=<browserstack-username>
BROWSERSTACK_KEY=<browserstack-access-key>
````

In the process, I found [one issue](https://travis-ci.org/shirish87/node-browserstack/jobs/106706376#L228) with `client.getLatest()` method; will give a separate PR for it.

Let me know if you need any changes.

Thanks!

### CODE REVIEW

**Comment by scottgonzalez** [comment]
Thanks. I'll probably need a few days before I get a chance to look at this. In the meantime, can you fix the build errors?

**Comment by shirish87** [comment]
No problem.
The build error is due to env variables for the BrowserStack account not being set.

Since this doesn't mock APIs but makes actual API calls, this would need an active BrowserStack account for CI, which I've now added. :)

File: test/api.js [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+
+				var requests = browsers.length;
+
+				browsers.forEach(function(browser) {
+					client.getLatest(browser, function(err, version) {
+						should.ifError(err);
+						version.should.match(/\d+(\.)*\d*/);
+
+						if (err || --requests < 1) {
+							done && done(err);
```
Comment (scottgonzalez) [inline test/api.js]: I don't like conditional guards like this. Please change this to:

```js
if (done) {
    done(err);
    done = null;
}
```

File: .jshintrc [hunk 2]
```diff
@@ -17,5 +17,12 @@
...
 		"define": false,
 		"PointerEvent": true
-	}
+	},
+
+	"predef": [
+		"describe",
+		"it",
+		"beforeEach",
+		"afterEach"
```
Comment (scottgonzalez) [inline .jshintrc]: Please move these to `globals` with a value of `false`.

File: test/client.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+			var client = BrowserStack.createClient({
+				username: username,
+				password: password
+			});
+
+			assert.equal(className, client.constructor.name, 'instance of ' + className);
+		});
+
+		it('should return an API client of a particular version', function() {
+			for (var v = 1; v <= apiLatestVersion; v++) {
```
Comment (scottgonzalez) [inline test/client.js]: `v` -> `version`

File: test/client.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+		it('should throw an error for invalid ApiClient version', function() {
+			try {
+				BrowserStack.createClient({
+					username: username,
+					password: password,
+					version: -1
+				});
+			} catch (e) {
+				assert.ok(e.toString().match(/invalid version/i));
```
Comment (scottgonzalez) [inline test/client.js]: `e` -> `err`

File: test/client.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+			}
+		});
+
+		it('should contain public methods', function() {
```
Comment (scottgonzalez) [inline test/client.js]: Since we have tests for the individual methods, this can be removed.

File: test/client.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		it('should return a screenshot client', function() {
+			var client = BrowserStack.createScreenshotClient({
+				username: username,
+				password: password
+			});
+
+			assert.equal(className, client.constructor.name, 'instance of ' + className);
+		});
+
+		it('should contain public methods', function() {
```
Comment (scottgonzalez) [inline test/client.js]: Same here. Even without tests, if a public method somehow disappeared, we'd know pretty quickly.

File: test/util.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		client.getJob.bind(client),
+		function getWorkerId(worker) {
+			return worker && (worker.job_id || worker.id);
+		},
+		function isWorkerRunning(worker) {
+			return worker && worker.state === 'running';
+		}, callback);
+};
+
+module.exports.validateBrowserObject = function validateBrowserObject(b) {
```
Comment (scottgonzalez) [inline test/util.js]: `b` -> `browser`

File: test/api.js [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+				versions.should.be.an.Object().and.not.be.empty();
+				done(err);
+			});
+		});
+
+		it('should get the latest version for specified browser', function(done) {
+			client.getBrowsers(function(err, browsers) {
+				should.ifError(err);
+
+				browsers = browsers.filter(function(b) {
```
Comment (scottgonzalez) [inline test/api.js]: `b` -> `browser`

**Comment by scottgonzalez** [comment]
Overall this looks good. The Travis build did fail though; it seems like it should have passed, so I'm going to restart the build.

**Comment by scottgonzalez** [comment]
Well, now the build fails because it runs against a merge and we added linting via JSCS. Can you rebase this, run `npm run jscs-fix` and push your updates?

File: test/client.js (hunk 6 above)
Comment (kartsims) [inline test/client.js]: Not sure all public methods are tested, `getJob` is not for instance. But maybe it should

Comment (scottgonzalez) [inline test/client.js]: Let's add a test for `getJob()` and drop these tests. The only other method that isn't tested in `ApiClient#getBrowserId()` but that shouldn't have been in the list anyway.

**Comment by shirish87** [comment]
Updated as per your comments + rebased + squashed. :)

**Comment by scottgonzalez** [comment]
Thanks, this has been merged!

I did make one small change: The environment variables are only expected to be base64 encoded in Travis. For local runs, they're just plaintext. See https://github.com/scottgonzalez/node-browserstack/blob/f66d857134b0be55f6a03d8b7157a7bac381a664/test/util.js#L8-L12 for the minor change.

**Pull request closed without merging**